module_3/gift_certificates_system/model/src/main/java/com/epam/esm/entity/GiftCertificateToTagRelation.java

```diff
@@ -0,0 +1,49 @@
+package com.epam.esm.entity;
+
+import com.epam.esm.listener.GiftCertificateToTagRelationListener;
+import com.epam.esm.util.DatabaseColumnName;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+import javax.persistence.Entity;
+import javax.persistence.EntityListeners;
+import javax.persistence.Id;
+import javax.persistence.JoinColumn;
+import javax.persistence.ManyToOne;
+import java.io.Serializable;
+import java.util.Objects;
+
+
+/**
+ * Gift certificate to tag relation entity.
+ */
+@AllArgsConstructor
+@NoArgsConstructor
+@Entity
+@Getter
+@EntityListeners(GiftCertificateToTagRelationListener.class)
+public class GiftCertificateToTagRelation implements Serializable {
+    @Id
+    @ManyToOne
+    @JoinColumn(name = DatabaseColumnName.GIFT_CERTIFICATE_ID)
+    private GiftCertificate giftCertificate;
+
+    @Id
+    @ManyToOne
+    @JoinColumn(name = DatabaseColumnName.TAG_ID)
+    private Tag tag;
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        GiftCertificateToTagRelation that = (GiftCertificateToTagRelation) o;
+        return giftCertificate.equals(that.giftCertificate) && tag.equals(that.tag);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(giftCertificate, tag);
+    }
```

Comment:
Why didn't you define this via lombok?

Reply:
Using `@EqualsAndHashCode` for JPA entities is not recommended. It can cause severe performance and memory consumption issues.